### ISSUE TITLE
Panic in RPC HTTP handler

### ISSUE THREAD

**Issue by Carameleon** [tendermint/tendermint#5069]
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):
```
# This is a part of go.mod in gaia
github.com/tendermint/go-amino v0.15.1
github.com/tendermint/tendermint v0.32.10
github.com/tendermint/tm-db v0.2.0
```

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
gaiad version --long
```
name: gaia
server_name: gaiad
client_name: gaiacli
version: 2.0.8
commit: e6f3efff002f74741b05233fe1f05aa893611d53
build_tags: netgo,ledger
go: go version go1.14.1 linux/amd64
```

**Environment**:
- **OS** (e.g. from /etc/os-release): 
OS version
```
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- **Install tools**:
- **Others**:

**What happened**:
While operating our full node and rest server for our applications at Cosmostation, we encounter the following error 3-4 times a day recently (we encountered this from time to time, but not this many), and our full node stops to sync. We are trying to find the root cause of this to resolve the issue. Do you have any idea why this panic occurs? Any suggestion?

**What you expected to happen**:

**Have you tried the latest version**: no

**How to reproduce it** (as minimally and precisely as possible): Not too sure about how to reproduce. It just occurs from time to time, but recently we encounter this 3-4 times a day.

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:
```
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: E[2020-07-01|02:09:00.372] Panic in RPC HTTP handler                    module=rpc-server err="write tcp xxx.xxx.xxx.xxx:26657->xxx.xxx.xxx.xxx:48892: i/o timeout" stack="goroutine 3603520009 [running]:\nruntime/debug.Stack(0xc00268d260, 0x105d500, 0xc00eb53310)\n\t/snap/go/5569/src/runtime/debug/stack.go:24 +0x9d\ngithub.com/tendermint/tendermint/rpc/lib/server.RecoverAndLogHandler.func1.1(0xc013dec680, 0x154c700, 0xc0025261a0, 0xbfb719ea757016fb, 0x9d0db210ef138, 0x1f8df40, 0xc014eae800)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:161 +0x486\npanic(0x105d500, 0xc00eb53310)\n\t/snap/go/5569/src/runtime/panic.go:967 +0x166\ngithub.com/tendermint/tendermint/rpc/lib/server.WriteRPCResponseHTTP(0x1546e80, 0xc013dec680, 0x116c06d, 0x3, 0x152fd60, 0xc013669450, 0xc0059f8000, 0x6771, 0x69c8, 0x0)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:112 +0x281\ngithub.com/tendermint/tendermint/rpc/lib/server.WriteRPCResponseArrayHTTP(0x1546e80, 0xc013dec680, 0xc00ec65000, 0x1, 0x1)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:121 +0xca\ngithub.com/tendermint/tendermint/rpc/lib/server.makeJSONRPCHandler.func1(0x1546e80, 0xc013dec680, 0xc014eae800)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/handlers.go:190 +0x49a\ngithub.com/tendermint/tendermint/rpc/lib/server.handleInvalidJSONRPCPaths.func1(0x1546e80, 0xc013dec680, 0xc014eae800)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/handlers.go:204 +0x9c\nnet/http.HandlerFunc.ServeHTTP(0xc00001a7f0, 0x1546e80, 0xc013dec680, 0xc014eae800)\n\t/snap/go/5569/src/net/http/server.go:2012 +0x44\nnet/http.(*ServeMux).ServeHTTP(0xc003336340, 0x1546e80, 0xc013dec680, 0xc014eae800)\n\t/snap/go/5569/src/net/http/server.go:2387 +0x1a5\ngithub.com/tendermint/tendermint/rpc/lib/server.maxBytesHandler.ServeHTTP(0x152e740, 0xc003336340, 0xf4240, 0x1546e80, 0xc013dec680, 0xc014eae800)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:210 +0xcf\ngithub.com/tendermint/tendermint/rpc/lib/server.RecoverAndLogHandler.func1(0x1547580, 0xc00031ae00, 0xc014eae800)\n\t/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:183 +0x360\nnet/http.HandlerFunc.ServeHTTP(0xc001cb8900, 0x1547580, 0xc00031ae00, 0xc014eae800)\n\t/snap/go/5569/src/net/http/server.go:2012 +0x44\nnet/http.serverHandler.ServeHTTP(0xc00031a380, 0x1547580, 0xc00031ae00, 0xc014eae800)\n\t/snap/go/5569/src/net/http/server.go:2807 +0xa3\nnet/http.(*conn).serve(0xc0065b65a0, 0x154bd00, 0xc00e582840)\n\t/snap/go/5569/src/net/http/server.go:1895 +0x86c\ncreated by net/http.(*Server).Serve\n\t/snap/go/5569/src/net/http/server.go:2933 +0x35c\n"
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: 2020/07/01 02:09:00 http: superfluous response.WriteHeader call from github.com/tendermint/tendermint/rpc/lib/server.(*ResponseWriterWrapper).WriteHeader (http_server.go:195)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: 2020/07/01 02:09:00 http: panic serving xxx.xxx.xxx.xxx:48892: write tcp xxx.xxx.xxx.xxx:26657->xxx.xxx.xxx.xxx:48892: i/o timeout
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: goroutine 3603520009 [running]:
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: net/http.(*conn).serve.func1(0xc0065b65a0)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:1772 +0x139
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: panic(0x105d500, 0xc00eb53310)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/runtime/panic.go:973 +0x3e3
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.WriteRPCResponseHTTPError(0x1546e80, 0xc013dec680, 0x1f4, 0x116c06d, 0x3, 0x152fd60, 0x1512670, 0x0, 0x0, 0x0, ...)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:100 +0x285
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.RecoverAndLogHandler.func1.1(0xc013dec680, 0x154c700, 0xc0025261a0, 0xbfb719ea757016fb, 0x9d0db210ef138, 0x1f8df40, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:163 +0x672
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: panic(0x105d500, 0xc00eb53310)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/runtime/panic.go:967 +0x166
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.WriteRPCResponseHTTP(0x1546e80, 0xc013dec680, 0x116c06d, 0x3, 0x152fd60, 0xc013669450, 0xc0059f8000,
 0x6771, 0x69c8, 0x0)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:112 +0x281
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.WriteRPCResponseArrayHTTP(0x1546e80, 0xc013dec680, 0xc00ec65000, 0x1, 0x1)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:121 +0xca
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.makeJSONRPCHandler.func1(0x1546e80, 0xc013dec680, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/handlers.go:190 +0x49a
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.handleInvalidJSONRPCPaths.func1(0x1546e80, 0xc013dec680, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/handlers.go:204 +0x9c
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: net/http.HandlerFunc.ServeHTTP(0xc00001a7f0, 0x1546e80, 0xc013dec680, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:2012 +0x44
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: net/http.(*ServeMux).ServeHTTP(0xc003336340, 0x1546e80, 0xc013dec680, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:2387 +0x1a5
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.maxBytesHandler.ServeHTTP(0x152e740, 0xc003336340, 0xf4240, 0x1546e80, 0xc013dec680, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:210 +0xcf
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: github.com/tendermint/tendermint/rpc/lib/server.RecoverAndLogHandler.func1(0x1547580, 0xc00031ae00, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/home/ubuntu/go/pkg/mod/github.com/tendermint/tendermint@v0.32.10/rpc/lib/server/http_server.go:183 +0x360
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: net/http.HandlerFunc.ServeHTTP(0xc001cb8900, 0x1547580, 0xc00031ae00, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:2012 +0x44
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: net/http.serverHandler.ServeHTTP(0xc00031a380, 0x1547580, 0xc00031ae00, 0xc014eae800)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:2807 +0xa3
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: net/http.(*conn).serve(0xc0065b65a0, 0x154bd00, 0xc00e582840)
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:1895 +0x86c
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: created by net/http.(*Server).Serve
Jul  1 02:09:00 ip-10-0-0-199 gaiad[886]: #011/snap/go/5569/src/net/http/server.go:2933 +0x35c
```

**Config (you can paste only the changes you've made)**:

**node command runtime flags**:
gaiad is registered in system service. Run command is `sudo systemctl start gaiad`
```
$ ps -ef | grep gaiad
root       886     1 99 May30 ?        35-13:47:31 /home/ubuntu/go/bin/gaiad start --home /home/ubuntu/.gaiad
```

**`/dump_consensus_state` output for consensus bugs**

**Anything else we need to know**:
I am not sure if this may help you to find the cause. CPU usage spiked up to 60 - 70% (sometimes up to 90%) when that our full node encountered that panic, and I discovered that over 100,000 times context switching occurred in 30 seconds. If you want me to provide any type of data, feel free to let me know.
```
part of shell script 
...
c_vol=`cat /proc/<gaiad_pid>/status | grep -w voluntary_ctxt_switches | awk '{print $2}'`
 c_nonvol=`cat /proc/<gaiad_pid>/status | grep -w nonvoluntary_ctxt_switches | awk '{print $2}'`
let diff_vol=c_vol-vol["$2"]
let diff_nonvol=c_nonvol-nonvol["$2"]
echo "diff vol_ctxt = ${diff_vol}"
echo "diff nonvol_ctxt = ${diff_nonvol}"
...
=================================
gaiad
=================================
diff vol_ctxt = 64026
diff nonvol_ctxt = 1542
Wed Jun  3 10:07:20 UTC 2020
=================================
gaiad
=================================
diff vol_ctxt = 100924
diff nonvol_ctxt = 2114
Wed Jun  3 10:07:50 UTC 2020
=================================
gaiad
=================================
diff vol_ctxt = 100070
diff nonvol_ctxt = 1498
Wed Jun  3 10:08:20 UTC 2020
```

**Comment by melekes** [comment]
The cause is whomever calling your RPC and terminating connection before receiving the response. We've fixed the panic itself in https://github.com/tendermint/tendermint/commit/413e554fd0f0fd87284a2a786ec866fcbefd19fe (v0.33.5), BUT it's okay for HTTP server to panic. 

What is not okay is for a full node to stop syncing. Are there any other logs besides panics before node stops syncing? Could you provide the goroutine dump (`kill -6 PID` or https://docs.tendermint.com/master/tools/debugging.html)?

**Comment by Carameleon** [comment]
> The cause is whomever calling your RPC and terminating connection before receiving the response. We've fixed the panic itself in [413e554](https://github.com/tendermint/tendermint/commit/413e554fd0f0fd87284a2a786ec866fcbefd19fe) (v0.33.5), BUT **it's okay** for HTTP server to panic.
> 
> **What is not okay** is for a full node to stop syncing. Are there any other logs besides panics before node stops syncing? Could you provide the goroutine dump (`kill -6 PID` or https://docs.tendermint.com/master/tools/debugging.html)?

It turns out my mistake about writing prometheus query.
I found our node sync at least 4 blocks per 1 minute at the panic or some state.
So, I think It has no problem any more.
If you think that is problem, I gonna try to offer additional information you mentioned.
Sorry for reporting wrong issue.

**Comment by melekes** [comment]
No worries! I'll close the issue then.

Also, take a look at the upcoming state-sync feature https://docs.tendermint.com/master/spec/abci/apps.html#state-sync. It drastically reduces the time to spin up a new node / sync an existing one.
